hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// TODO: This is conceptually kinda close to ReferenceGraph except more powerful
+
+// typeWalkerRemoveType is a special TypeName that informs the type walker to remove the property containing this TypeName
+// entirely.
+var typeWalkerRemoveType = MakeTypeName(LocalPackageReference{}, "TypeWalkerRemoveProperty")
+
+// TODO: it's awkward to have so much configuration on this thing (3 separate funcs that apply at different places in the walking process?)
+// TODO: but unsure if there's a better way... bring it up in code review.
+
+// TypeWalker performs a depth first traversal across the types provided, applying the visitor to each TypeDefinition.
+// MakeContext is called before each visit, and AfterVisit is called after each visit. RemoveCycle is called
+// if a cycle is detected.
+type TypeWalker struct {
+	allTypes Types
+	visitor  TypeVisitor
+
+	// MakeContext is called before any type but the root type is visited. It is given the current context and returns
+	// a new context for use in the upcoming Visit.
+	MakeContext func(it TypeName, ctx interface{}) (interface{}, error)
+	// AfterVisit is called after the type walker has applied the visitor to a TypeDefinition.
+	AfterVisit func(original TypeDefinition, updated TypeDefinition, ctx interface{}) (TypeDefinition, error)
```

Comment:
Now I really want these to be called `PushContext` and `PopContext` (or `PushVisit` and `PopVisit`?). I think those make sense with how they're used, even the nil being passed in for the initial context (like the next pointer in a linked list being used as a stack).

The MakeContext doc needs updating now that it's also called for the root type.

Reply:
Fixed